exec/java-exec/src/main/java/org/apache/drill/exec/store/TimedCallable.java

```diff
@@ -0,0 +1,265 @@
+      } else {
+        logger.info("{}: completed {} out of {} using {} threads", activity, doneCount, count, parallelism);
+      }
+    }
+  }
+
+  @Override
+  public final V call() throws Exception {
+    long start = System.nanoTime();
+    startTime = start;
+    try {
+      logger.debug("Started execution of '{}' task at {} ms", this, TimeUnit.MILLISECONDS.convert(start, TimeUnit.NANOSECONDS));
+      return runInner();
+    } catch (InterruptedException e) {
+      logger.warn("Task '{}' interrupted", this, e);
+      throw e;
+    } finally {
+      long time = System.nanoTime() - start;
+      if (logger.isWarnEnabled()) {
+        long timeMillis = TimeUnit.MILLISECONDS.convert(time, TimeUnit.NANOSECONDS);
+        if (timeMillis > TIMEOUT_PER_RUNNABLE_IN_MSECS) {
+          logger.warn("Task '{}' execution time {} ms exceeds timeout {} ms.", this, timeMillis, TIMEOUT_PER_RUNNABLE_IN_MSECS);
+        } else {
+          logger.debug("Task '{}' execution time is {} ms", this, timeMillis);
+        }
+      }
+      executionTime = time;
+    }
+  }
+
+  protected abstract V runInner() throws Exception;
+
+  private long getStartTime(TimeUnit unit) {
+    return unit.convert(startTime, TimeUnit.NANOSECONDS);
+  }
+
+  private long getExecutionTime(TimeUnit unit) {
+    return unit.convert(executionTime, TimeUnit.NANOSECONDS);
+  }
+
+
+  /**
+   * Execute the list of runnables with the given parallelization.  At end, return values and report completion time
+   * stats to provided logger. Each runnable is allowed a certain timeout. If the timeout exceeds, existing/pending
+   * tasks will be cancelled and a {@link UserException} is thrown.
+   * @param activity Name of activity for reporting in logger.
+   * @param logger The logger to use to report results.
+   * @param tasks List of callable that should be executed and timed.  If this list has one item, task will be
+   *                  completed in-thread. Each callable must handle {@link InterruptedException}s.
+   * @param parallelism  The number of threads that should be run to complete this task.
+   * @return The list of outcome objects.
+   * @throws IOException All exceptions are coerced to IOException since this was build for storage system tasks initially.
+   */
+  public static <V> List<V> run(final String activity, final Logger logger, final List<TimedCallable<V>> tasks, int parallelism) throws IOException {
+    Preconditions.checkArgument(!Preconditions.checkNotNull(tasks).isEmpty(), "list of tasks is empty");
+    parallelism = Math.min(parallelism, tasks.size());
+    final ExecutorService threadPool = parallelism == 1 ? MoreExecutors.newDirectExecutorService()
+        : Executors.newFixedThreadPool(parallelism, new ThreadFactoryBuilder().setNameFormat(activity + "-%d").build());
+    final long timeout = (long)Math.ceil((TIMEOUT_PER_RUNNABLE_IN_MSECS * tasks.size())/parallelism);
```

Comment:
IMO, this code part of the original logic needs to be changed to TIMEOUT_PER_RUNNABLE_IN_MSECS * (long)Math.ceil(tasks.size()/parallelism); instead of (long)Math.ceil((TIMEOUT_PER_RUNNABLE_IN_MSECS * tasks.size())/parallelism).